Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -0,0 +1,337 @@
+        case topSites
+        case history
+
+        static let count = 2
+
+        var title: String? {
+            switch self {
+            case .history: return "Recent Activity"
+            default: return nil
+            }
+        }
+
+        var headerHeight: CGFloat {
+            switch self {
+            case .history: return 40
+            default: return 0
+            }
+        }
+
+        var headerView: UIView? {
+            switch self {
+            case .history:
+                let view = ASHeaderView()
+                view.title = "Recent Activity"
+                return view
+            default:
+                return nil
+            }
+        }
+
+        var cellIdentifier: String {
+            switch self {
+            case .topSites: return "TopSite"
+            default: return "Cell"
+            }
+        }
+
+        init(at indexPath: NSIndexPath) {
+            self.init(rawValue: indexPath.section)!
+        }
+
+        init(_ section: Int) {
+            self.init(rawValue: section)!
+        }
+    }
+
+}
+
+// MARK: -  Header Views
+extension ActivityStreamPanel {
+
+    func tableView(tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        return Section(section).headerHeight
+    }
+
+    func tableView(tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        return Section(section).headerView
+    }
+}
+
+// MARK: - Tableview management
+extension ActivityStreamPanel: UITableViewDelegate, UITableViewDataSource {
+
+    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
+        return Section.count
+    }
+
+    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        switch Section(section) {
+            case .topSites:
+                if !topSiteHandler.content.isEmpty {
+                    return 1
+                } else {
+                    return 0
+                }
+            case .history:
+                 return self.history.count
+        }
+    }
+
+    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let identifier = Section(indexPath.section).cellIdentifier
+        let cell = tableView.dequeueReusableCellWithIdentifier(identifier, forIndexPath: indexPath)
+
+        switch Section(indexPath.section) {
+            case .topSites:
+                return configureTopSitesCell(cell, forIndexPath: indexPath)
+            default:
+                return configureHistoryItemCell(cell, forIndexPath: indexPath)
+        }
+    }
+
+    func configureTopSitesCell(cell: UITableViewCell, forIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let topSiteCell = cell as! ASHorizontalScrollCell
+        // The topSiteCell needs a refrence to the tableview because it needs to alert the tableView to relayout once topsites finishes loading.
+        topSiteCell.parentTableView = self.tableView
```

Comment:
This is one of my least favorite hacks. there was no way I could think of to relayout the tableview without adding this reference 😢